bpo-43605: Improve the documentation to exec() and eval() (#25039)
python/cpython: Doc/library/functions.rst

.. function:: eval(expression[, globals[, locals]])

>    This function supports the dynamic evaluation of Python expression.  The

Reviewer: ```suggestion
   This function supports the dynamic evaluation of Python expressions.  The
```

Author: Good catch! Thank you!

   first argument can be a string or a code object.  The optional arguments
   specify the globals and locals respectively.  If provided, *globals* must be

>    a dictionary.  If provided, *locals* can be any mapping object.

Reviewer: I don't think `locals` can be any mapping object. Eg. ``collections.Counter()`` is a mapping object (``isinstance(collections.Counter('a'), collections.abc.Mapping)`` is ``True``), but it cannot be used for eval:
```python
>>> Counter("abcd")
Counter({'a': 1, 'b': 1, 'c': 1, 'd': 1})
>>> eval("print(a)", {}, Counter("abcd"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 1, in <module>
TypeError: 'int' object is not callable
>>> eval("print(a)", {}, {'a':1})
1
```

Surprisingly, git blame says this line has been around since this paragraph's inception 14 years ago.

```suggestion
   specify the globals and locals respectively.  If provided, *globals*
   and *locals* must be dictionaries.
```

Author: What?? In the source for `eval/exec` in Python/bltinmodule.c, the locals are checked by `PyMapping_Check()` while globals checked by `PyDict_Check()`. I'm not sure why the exception was about "'int' object is not callable" -- it seems the name `a` did get resolved in the locals through the `Counter` object, but somehow there was an action to the effect of `1()`. <s>This might be a bug! Right now I don't have enough spoons to track it down, but I should be verifying it and maybe submitting a bug report soon.</s> Edit: working as intended, but in a tricky way: see [my reply below](https://github.com/python/cpython/pull/25039#discussion_r602783224).

Author: This works as intended:
```python
>>> ct = Counter("abcd")
>>> eval("a", {}, ct)
1
```
But not this:
```python
>>> eval("print(a)", {}, ct)
```
Expected return value of `eval` is `None` (return value of `print` function), but the cryptic exception is raised.

Author: Also, this:
```python
>>> eval("print(1)", {}, ct)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 1, in <module>
TypeError: 'int' object is not callable
```

Author: I see! The locals have the highest priority in name resolution. The `Counter` instance (`ct` in the above example) behaves as if any key is in it (with initial value 0). When the machinery underlying `eval` does its work, it first looks up the name `"print"` in the locals given to it. Unfortunately, as a result of how `Counter` behaves, it gets what it wishes for: the count for the key `"print"` in the counter `ct`, which is the `int` instance `0`. Then `0` gets called as the function.  The evaluation boils down to `0(1)` which raises `TypeError`.

Reviewer: Awesome! Thank you for investigating this! I learnt a lot :).

I forgot that Counter sort of behaves like a defaultdict and returns 0 for keys it doesn't have.

Author: Thanks, me too. This is the weirdest thing I've seen in Python in a while.

   In the most common case, the *expression* argument is a string, and it is
   parsed and evaluated as a Python expression (see the section

>    ":ref:`expression-input`" in the Language Reference).  The leading and

Reviewer: Sorry, I don't remember if this is valid/invalid inline markup. Some roles require whitespace around them. See this to learn more:
https://devguide.python.org/documenting/#inline-markup

So it might be ``"\ :ref:`expression-input`\ "``. However, if you ran ``./make html`` and all looks fine on your computer, then I'm wrong and please ignore this comment :).

Author: Hi, I think this works fine (verified with the Sphinx setup created by "make venv" and "make html") -- unlike inline markup, the `ref` cross-linking markup seem to be Ok near punctuations.

.. function:: eval(expression[, globals[, locals]])

>    This function supports the dynamic evaluation of Python expressions.  The

Reviewer 2: ```suggestion
   This function dynamically evaluates Python expressions.  The
```

>    first argument can be a string or a code object.  The optional arguments

Reviewer 2: Maybe link this to the documentation for code objects.

   specify the global and local namespaces respectively.  If provided,
   *globals* must be a dictionary.  If provided, *locals* can be any
   :term:`mapping` object.

>    In the most common case, the *expression* argument is a string, and it is

Reviewer 2: ```suggestion
   When the *expression* argument is a string, it is
```
Tighten the wording

   parsed and evaluated as a Python expression (see the section
   ":ref:`expression-input`" in the Language Reference).  The leading and
   trailing spaces, tabs, and newlines are stripped.

   The evaluation is performed in the environment specified by the arguments
   *globals* and *locals*.  If both are omitted, by default it uses the
   environment where :func:`eval` is called.  If only the *globals* argument is
   given, the local namespace defaults to *globals*.

   Before evaluation, the special key ``"__builtins__"`` is searched for in the

>    global namespace that is explicitly or impcitly specified for :func:`eval`.

Reviewer 2: ```suggestion
   global namespace that is explicitly or implicitly specified.
```

      Raises an :ref:`auditing event <auditing>` ``exec`` with the code object
      as the argument. Code compilation events may also be raised.

   .. _eval_limitation_dynamic:
   .. note::

      Dynamic evaluation at run-time is not equivalent to embedding the
      expression at the same place in a Python program and having it compiled
      as a part of the whole program; this is explained in the section
      ":ref:`dynamic-features`" in the Language Reference.

      In particular, :func:`eval` does not have access to the :term:`nested
      scopes <nested scope>` (non-locals) in the enclosing environment.  Of
      note are expressions such as :term:`lambdas <lambda>`,
      :ref:`comprehensions <comprehensions>`, and :term:`generator expressions
      <generator expression>` which create an inner scope of their own.  Since
      Python 3.8, the action of an assignment expression (see :pep:`572`) also

>       depends on scope information determined at compile-time.  Interaction

Reviewer 2: ```suggestion
      depends on scope information determined at compile time.  Interaction
```
